File: dp/SubsetSum.java
public class SubsetSum{

    public boolean canPartition(int num[]){
        int sum = 0;

        for(int i=0; i<num.length ; i++)
            sum += num[i];

        if(sum % 2 != 0)
            return false;

        Boolean dp[][] = new Boolean[num.length][sum/2 + 1];
        return this.partition(dp,num , sum/2 , 0);
    }

    public boolean partition(Boolean dp[][], int num[] , int sum, int index){
        if(sum == 0)
            return true;
        
        if(num.length == 0 || index >= num.length)
            return false;


        if(dp[index][sum] == null){
            if(num[index] <= sum){
                if(partition(dp,num, sum - num[index], index + 1)){
                    dp[index][sum] = true;
                    return true;
                }
              }
            //  else{
              dp[index][sum] = partition(dp, num, sum, index+1);
            //  }
               
        }
        return dp[index][sum];

    }

    public static void main(String args[]){
        SubsetSum s = new SubsetSum();
        int num[] = {1,2,3,8};
        System.out.println(s.canPartition(num));
    }
}

File: dp/subsetsum.cpp
// recursive solution to find subset partition(Top down approach)
#include<bits/stdc++.h>
using namespace std;

bool dp[10][10];

int  issubset( int arr[], int n, int sum)
{
    if(sum == 0 )
        return true;
    if(n ==0 && sum != 0)
        return false;

     if(dp[n-1][sum] == true)
        return dp[n-1][sum];
        

    if(arr[n-1] > sum)
        return  dp[n-1][sum] = issubset(arr,n-1,sum);

    return (dp[n-1][sum] = (issubset(arr,n-1,sum-arr[n-1]) || issubset(arr,n-1,sum)) );
}



int main()
{
    
    cout<<dp[3][3]<<'\n';
    int arr[] = {2,2,2};
    int n = sizeof(arr)/sizeof(arr[0]);
    int sum=0;
    for(int i=0 ; i<n;i++)
        sum += arr[i];

    if(sum % 2 != 0)
    {
        cout<<"false\n";
        return 0;
    }

    
    cout<<dp[3][3]<<'\n';

    bool b = issubset(arr,n,sum/2) ;

    for(int i=0 ; i<3; i++)
    {
        for(int j=0 ; j<=8 ; j++)
        {
            cout<<dp[i][j]<<" ";
        }
        cout<<endl;
    }

    
    if(b == false)
        cout<<"No\n";
    else 
    {
        cout<<"Yes\n";
    }
    
    return 0;
}